packages/office-ui-fabric-react/src/components/Coachmark/Coachmark.tsx

```diff
@@ -165,56 +239,161 @@ export class Coachmark extends BaseComponent<ICoachmarkTypes, ICoachmarkState> {
       this._async.setTimeout(() => {
-        this._addProximityHandler(100);
+        this._addProximityHandler(this.props.mouseProximityOffset);
       }, this.props.delayBeforeMouseOpen!);
     }));
   }
 
   private _onFocusHandler = (): void => {
-    this._openCoachmark();
+    if (this.state.isCollapsed) {
+      this._openCoachmark();
+    }
   }
 
-  private _openCoachmark = (): void => {
+  private _onPositioned = (positionData: IPositionedData): void => {
+    this._async.requestAnimationFrame(((): void => {
+      this.setState({
+        targetAlignment: positionData.alignmentEdge,
+        targetPosition: positionData.targetEdge
+      });
+    }));
+  }
+
+  private _getBounds(): IRectangle | undefined {
+    const { isPositionForced } = this.props;
+    if (isPositionForced) {
+      return {
+        left: -Infinity,
+        top: -Infinity,
+        bottom: Infinity,
+        right: Infinity,
+        width: Infinity,
+        height: Infinity
+      };
+    } else {
+      return undefined;
+    }
+  }
+
+  private _setBeakPosition = (): void => {
+    let beakLeft;
+    let beakTop;
+    let beakRight;
+    let beakBottom;
+    let transformOriginX;
+    let transformOriginY;
+
+    const { targetAlignment } = this.state;
+    const distanceAdjustment = '3px';  // Adjustment distance for Beak to shift towards Coachmark bubble.
+
+    switch (this._beakDirection) {
+      // If Beak is pointing Up or Down
+      case BeakDirection.Top:
+      case BeakDirection.Bottom:
+        // If there is no target alignment, then beak is X-axis centered in callout
+        if (!targetAlignment) {
+          beakLeft = `calc(50% - ${BEAK_WIDTH / 2}px)`;
+          transformOriginX = 'center';
+        } else {
+          // Beak is aligned to the left of target
+          if (targetAlignment === RectangleEdge.left) {
+            beakLeft = `calc(${COACHMARK_WIDTH / 2}px - ${BEAK_WIDTH / 2}px)`;
+            transformOriginX = 'left';
+          } else {
+            // Beak is aligned to the right of target
+            beakRight = `calc(${COACHMARK_WIDTH / 2}px - ${BEAK_WIDTH / 2}px)`;
+            transformOriginX = 'right';
+          }
+        }
+
+        if (this._beakDirection === BeakDirection.Top) {
+          beakTop = distanceAdjustment;
+          transformOriginY = 'top';
+        } else {
+          beakBottom = distanceAdjustment;
+          transformOriginY = 'bottom';
+        }
+        break;
+      // If Beak is pointing Left or Right
+      case BeakDirection.Left:
+      case BeakDirection.Right:
+        // If there is no target alignment, then beak is Y-axis centered in callout
+        if (!targetAlignment) {
+          beakTop = `calc(50% - ${BEAK_WIDTH / 2}px)`;
+          transformOriginY = `center`;
+        } else {
+          // Beak is aligned to the top of target
+          if (targetAlignment === RectangleEdge.top) {
+            beakTop = `calc(${COACHMARK_WIDTH / 2}px - ${BEAK_WIDTH / 2}px)`;
```

Comment:
Why are you using calc? That's likely going to be slower than just doing the calculation right here in javascript.